### PR TITLE
Next cut new backend

### PR DESCRIPTION
**Pull request by cliffclick**
Makefile builds javatools.jar, ecstasy.xtc and all the XDK XTC files, and XCC, and uses built XCC to compile all XTC files.
Rewrite AST tree passes; doType, unBox, rewrite, reBox, jcode.
Rewrite Elvis handling.
Support for unions.

### CODE REVIEW

File: manualTests/src/main/x/new_backend/.dir_locals.el
Comment (ggleyzer) [inline manualTests/src/main/x/new_backend/.dir_locals.el:1]: I think this file is only neccesary for your local dev environment. Please remove it from the submission.

File: tck/src/main/x/tck/union/Basic.x [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class Basic {
+
+    void run() {
+        basic();
+    }
+
+    // -----------------------------
+    void basic() {
+        typedef (String | Int) as StrInt;
+        StrInt id( StrInt x ) { return x; }
```
Comment (ggleyzer) [inline tck/src/main/x/tck/union/Basic.x:13]: Not a big thing, but the code style in this line is different from the rest of the tck. I'd prefer a consistent look.

File: tck/src/main/x/tck/union/Basic.x [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        StrInt x = id(7);
+        assert x==7;
```
Comment (ggleyzer) [inline tck/src/main/x/tck/union/Basic.x:15]: Again, completely trivial and tiny style issue (for spaces): the rest of the tck would look like that:

    assert x == 7;
    assert x != id("abc");
    assert x == "def";

File: bin/makedepends.sh
Comment (ggleyzer) [inline bin/makedepends.sh:1]: This file seems to be back-end specific. Do we need it in the public repo?

**Review by ggleyzer — APPROVED** [review]
I have only looked at the non-backend-specific files. Please see my comments and fix if you agree

File: tck/src/main/x/tck/union/Basic.x (hunk 2 above)
Comment (cliffclick) [reply tck/src/main/x/tck/union/Basic.x:15]: changed

File: bin/makedepends.sh
Comment (cliffclick) [reply bin/makedepends.sh:1]: Basically its parallel to the top-level Makefile, and I put it next to the other top-level build scripts.
Is there a better place?

File: tck/src/main/x/tck/union/Basic.x (hunk 1 above)
Comment (cliffclick) [reply tck/src/main/x/tck/union/Basic.x:13]: I'm not sure what style you are asking for here?

Comment (ggleyzer) [reply tck/src/main/x/tck/union/Basic.x:13]: Either a long form:

    StrInt id(StrInt x) {
        return x;
    }    
or a short one:

    StrInt id(StrInt x) = x;

File: manualTests/src/main/x/new_backend/.dir_locals.el
Comment (cliffclick) [reply manualTests/src/main/x/new_backend/.dir_locals.el:1]: humm... I made "new_backend" specifically so I could edit TestMisc.x freely; the ".dir_locals.el" is basically an IdeaJ equiv project file in that directory saying "use 4 space indents".  I would like to have project files for my tooling (emacs), that say XTC-specific things and would travel with me to other machines.  The ideal place for this file (in my mind) is at the top-level like e.g. ".idea" is for IntelliJ.  Short of having it in the top-level, what's wrong with having it in a new_backend specific directory?

File: bin/makedepends.sh
Comment (ggleyzer) [reply bin/makedepends.sh:1]: Would it be possible to have javatools_backend/bin for both Makefile and makedepends.sh?

File: manualTests/src/main/x/new_backend/.dir_locals.el
Comment (ggleyzer) [reply manualTests/src/main/x/new_backend/.dir_locals.el:1]: You're right; I misread the location. No problem here

File: tck/src/main/x/tck/union/Basic.x (hunk 1 above)
Comment (cliffclick) [reply tck/src/main/x/tck/union/Basic.x]: done

File: bin/makedepends.sh
Comment (cliffclick) [reply bin/makedepends.sh:1]: All Things Are Possible, but very definitely not easy.  I'm using make to build xcc & all xtc files, as such it badly wants to be "above" them.  I can move makedepends to anyplace, but moving the makefile is a lot of grief.

Comment (ggleyzer) [reply bin/makedepends.sh:1]: K; let's keep the Makefile where it is and create javatools_backend/bin for everything else

Comment (cliffclick) [reply bin/makedepends.sh:1]: moved makedepends

**Pull request merged**